dbfit-java/derby/src/main/java/dbfit/environment/DerbyEnvironment.java

```diff
@@ -51,40 +52,199 @@ public Pattern getParameterPattern() {
+            return createDbParamAccessor(paramName, rs, getParamDirection(rs));
+        }
+    }
+
+    private class DatabaseProcedure extends DatabaseExecutable {
+
+        private DatabaseProcedure(DatabaseObjectName objName) {
+            super(objName);
+        }
+
+        @Override
+        protected ResultSet getObjectMetaData() throws SQLException {
+            return getDbMetaData().getProcedures(null, name.getSchemaName(), name.getObjectName());
+        }
+
+        @Override
+        public ResultSet getColumns() throws SQLException {
+            return getDbMetaData().getProcedureColumns(null, name.getSchemaName(), name.getObjectName(), "%");
+        }
+
+        @Override
+        protected Direction columnTypeToDir(int paramType) {
+            switch (paramType) {
+                case DatabaseMetaData.procedureColumnIn:
+                    return Direction.INPUT;
+                case DatabaseMetaData.procedureColumnInOut:
+                    return Direction.INPUT_OUTPUT;
+                case DatabaseMetaData.procedureColumnOut:
+                    return Direction.OUTPUT;
+                default:
+                    return null;
+            }
+        }
+    }
+
+    private class DatabaseFunction extends DatabaseExecutable {
+
+        private DatabaseFunction(DatabaseObjectName objName) {
+            super(objName);
+        }
+
+        @Override
+        protected ResultSet getObjectMetaData() throws SQLException {
+            return getDbMetaData().getFunctions(null, name.getSchemaName(), name.getObjectName());
+        }
+
+        @Override
+        public ResultSet getColumns() throws SQLException {
+            return getDbMetaData().getFunctionColumns(null, name.getSchemaName(), name.getObjectName(), "%");
+        }
+
+        @Override
+        protected Direction columnTypeToDir(int paramType) {
+            switch (paramType) {
+                case DatabaseMetaData.functionColumnIn:
+                    return Direction.INPUT;
+                case DatabaseMetaData.functionReturn:
+                    return Direction.RETURN_VALUE;
+                default:
+                    return null;
+            }
+        }
+    }
+
+    private class DatabaseObjectName {
+        String schemaName, objectName;
+
+        private DatabaseObjectName(String defaultSchemaName, String objName) {
+            String[] qualifiers = objName.split("\\.");
+            schemaName = (qualifiers.length == 2) ? qualifiers[0] : defaultSchemaName;
+            objectName = (qualifiers.length == 2) ? qualifiers[1] : qualifiers[0];
+        }
+
+        private String getSchemaName() {
+            return schemaName;
+        }
+
+        String getObjectName() {
+            return objectName;
+        }
+    }
+
+    DatabaseObjectName createDbObjectNameWithCurrentDefaultSchema(String objName) throws SQLException {
+        return new DatabaseObjectName(getConnection().getSchema(), objName);
```

Comment:
>> What about moving the parsing logic from DatabaseObjectName c-tor to here?
>
> What would that do for us?

* Would allow more expressive name for the `DatabaseObjectName` construction method.
* There is a general rant about factory methods vs constructors, and avoiding complex logic in constructors. As we already have a factory method (create/build ... with current default schema): it looks logical to keep non-core construction details outside the c-tor.
* It's interesting to have a look at [Kotlin Data Classes](https://kotlinlang.org/docs/reference/data-classes.html). Unfortunately not available in Java language itself (maybe some day).

> What about adding a parse method to the `DatabaseObjectName` class although it can't be static?

The issue is we can't use non-static method before construction... But it's possible to make `DatabaseObjectName` class static. Or the parsing method can be outside the class (as we now have buildDatabaseObjectName).